mesh shader culling

gpx1000 opened the pull request:
## Description

Adding a mesh sample that demonstrates basic culling.  Include a tutorial to describe basic ideas and link to best practices.

## General Checklist:

Please ensure the following points are checked:

- [x] My code follows the [coding style](https://github.com/KhronosGroup/Vulkan-Samples/tree/master/CONTRIBUTING.md#Code-Style)
- [x] I have reviewed file [licenses](https://github.com/KhronosGroup/Vulkan-Samples/tree/master/CONTRIBUTING.md#Copyright-Notice-and-License-Template)
- [x] I have updated the header Copyright to reflect the current year (CI build will fail if Copyright is out of date)
- [x] I have commented any added functions (in line with Doxygen)
- [x] I have commented any code that could be hard to understand
- [x] My changes do not add any new compiler warnings
- [x] My changes do not add any new validation layer errors or warnings
- [x] I have used existing framework/helper functions where possible
- [x] My changes build on Windows, Linux, macOS and Android. Otherwise I have [documented any exceptions](https://github.com/KhronosGroup/Vulkan-Samples/tree/master/CONTRIBUTING.md#General-Requirements)
- [x] My changes do not add any regressions
- [x] I have tested every sample to ensure everything runs correctly
- [x] This PR describes the scope and expected impact of the changes I am making

## Sample Checklist

If your PR contains a new or modified sample, these further checks must be carried out *in addition* to the General Checklist:
- [x] I have tested the sample on at least one compliant Vulkan implementation
- [x] If the sample is vendor-specific, I have [tagged it appropriately](https://github.com/KhronosGroup/Vulkan-Samples/tree/master/CONTRIBUTING.md#General-Requirements)
- [x] I have stated on what implementation the sample has been tested so that others can test on different implementations and platforms
- [x] Any dependent assets have been merged and published in downstream modules
- [x] For new samples, I have added a paragraph with a summary to the appropriate chapter in the [samples readme](./../samples/README.md)

wflohry committed:
Using basic raytracing as template

wflohry committed:
Load sponza

wflohry committed:
Loading sponza model

wflohry committed:
Separating out ray tracing command buffers

wflohry committed:
Add render modes

wflohry committed:
Edit vertex structure

wflohry committed:
Put static objects in buffer with device-exclusive memory

wflohry committed:
Adding vertex and index data to bindings

wflohry committed:
Demonstration of referencing model data from ray tracing shader

wflohry committed:
Implemented texture array and visualization

wflohry committed:
Added fire pit

wflohry committed:
Added ray-traced shadows

wflohry committed:
Adding support for AO

wflohry committed:
Updated model settings

gpx1000 committed:
Fixed a few warnings and spelling errors.

Also fixed the fences problem that causes crash bug from fences on close.

wflohry committed:
Separating out static model loading

wflohry committed:
Separate handling of static / dynamic buffers

wflohry committed:
Added dynamic object and refactored

gpx1000 committed:
Fixed another warning and fixed the fences.

gpx1000 committed:
Merge branch 'feature/raytrace-extended' into warn_removal_and_fence_fix

# Conflicts:
#	samples/extensions/raytracing_extended/raytracing_extended.cpp

gpx1000 committed:
Switch to using constants and specializations in order to change values.  Can switch back to uniforms at a later date but the uniforms were proving a render issue for linux.  This should also demonstrate how to use multiple specialization constants in a ray traced environment.

gpx1000 committed:
more warnings removal

wflohry committed:
Merge warn_remove_and_fence_fix

gpx1000 committed:
Merged in warn_removal_and_fence_fix (pull request #1)

Fixed a few warnings and spelling errors.

Approved-by: William Lohry

wflohry committed:
Merge branch 'feature/raytrace-extended' of ssh://bitbucket.org/holochip/vulkan-samples into feature/raytrace-extended

wflohry committed:
Moving instance buffer creation to TLAS construction

wflohry committed:
Adding refraction placeholder

wflohry committed:
Perform BLAS update for dynamic objects and TLAS update every frame.

wflohry committed:
Adding dynamic sinusoidal wave

wflohry committed:
Finished refractive wave

wflohry committed:
Update to refraction

wflohry committed:
Adding separate flame instance

wflohry committed:
Working flame / particle generator

wflohry committed:
Updates to flame and refraction calculation

wflohry committed:
Updated refraction algorithm

wflohry committed:
Added texture

gpx1000 committed:
more warnings removal
move the png to ktx format

gpx1000 committed:
ensure we're pointing to the right place

wflohry committed:
Merge and update

wflohry committed:
Update to ktx commit

gpx1000 committed:
run clang-format over the sample.

wflohry committed:
Added transformations

wflohry committed:
Merged

wflohry committed:
Merged

gpx1000 committed:
Merged in clang-format (pull request #3)

run clang-format over the sample.

Approved-by: William Lohry

wflohry committed:
Adding transformations

wflohry committed:
Implemented automatic  billboard rotation

wflohry committed:
Merge branch 'feature/raytrace-extended' of ssh://bitbucket.org/holochip/vulkan-samples into feature/raytrace-extended

gpx1000 committed:
run a profile pass.
move pts, uvs, normals, indices to be created/initialized once to avoid dynamic memory generation.
remove warnings.
Update copyright.

gpx1000 committed:
Merged in profile (pull request #4)

run a profile pass.

Approved-by: William Lohry

gpx1000 committed:
Update copyright in CMakeLists.txt
Update sample cmake function definition
Update samples list in Samples CMakeLists.txt
Update README.md with first pass definition of the sample.

gpx1000 committed:
Merged in draft_ready (pull request #5)

Update copyright in CMakeLists.txt

Approved-by: William Lohry

wflohry committed:
Added tutorial rough draft

gpx1000 committed:
grammar check.

gpx1000 committed:
grammar check.

gpx1000 committed:
Merged in tutorial (pull request #6)

grammar check.

Approved-by: William Lohry

wflohry committed:
Added fence reset

wflohry committed:
Updating triangle indexing code

wflohry committed:
Avoid dynamic allocation / copy

gpx1000 committed:
update copyright

gpx1000 committed:
Merged in feature/raytrace-extended (pull request #7)

Feature/raytrace extended

gpx1000 committed:
Merge branch 'downstream-master'

gpx1000 committed:
Fixes for Clang-format and copyright headers CI failures.

gpx1000 committed:
add in nonuniformEXT to the texture lookup as imageOffset isn't uniform.

gpx1000 committed:
switch to textureLod from texture to ensure it's ready for mipmaps if copied from sample.

wflohry committed:
Fixing refraction bug

gpx1000 committed:
Merge remote-tracking branch 'downstream/feature/raytrace-extended'

gpx1000 committed:
fix typo in README.md

gpx1000 committed:
fix typo in README.md and in CMakeLists.txt

gpx1000 committed:
Merge branch 'master' into master

wflohry committed:
Updates for PR comments

wflohry committed:
Adding memory barriers

wflohry committed:
Updating shaders to use defines

gpx1000 committed:
Merge remote-tracking branch 'downstream/feature/raytrace-extended'

# Conflicts:
#	samples/extensions/raytracing_extended/raytracing_extended.cpp
#	shaders/khr_ray_tracing_extended/closesthit.rchit

gpx1000 committed:
clang format cleanup

gpx1000 committed:
add missing copyright to new header

gpx1000 committed:
fix attribution in the tutorial.md

gpx1000 committed:
framework:
acceleration structure: (build) fix a bug which only would allow a single triangle to be included in a acceleration structure.
acceleration structure: (build) allow for more than one call to build (useful in animation).  This was caused by recreating the buffer each time rather than reusing it if it already exists.
acceleration structure: (add triangles) allow for developer to specify the data address of the vertex index and transform buffers that are used.  This allows for multi model gltf files to work as in the sponza scene.
acceleration structure: enable buffer retrieval to allow for developer to add in memory barriers.
acceleration structure: provide developer means to wipe the geometries added to the acceleration structure.  This is very important if animating and thus recalling the same structure(s).  NB: a future feature would be to index the geometries and instances to allow for updating existing data individually rather than a requirement to wipe out everything and rebuild everything.
scratch buffer: add a return for the scratch buffer's size to allow for reusing the scratch buffer if it's the same size as previously allocated.

Sample:
As this is the first ray tracing sample to make use of @Sascha William's excellent acceleration structure framework, and we already have the existing long form, I opted to enable a macro to turn on/off using the framework's acceleration structure.  I think this is valuable to end users to show a precise "what the frameork does" behind the scenes.  Future samples should opt to use the framework.

gpx1000 committed:
run clang-format over scratch_buffer.h

gpx1000 committed:
Add in proper update logic for the acceleration_structure in both the framework and the sample.

gpx1000 committed:
Ensure that only updated geometry are used in each update pass and make framework the default AS.

gpx1000 committed:
Merge branch 'master-upstream'

gpx1000 committed:
Small changes for buffer.

gpx1000 committed:
Merge branch 'master-upstream'

gpx1000 committed:
Merge branch 'master-upstream'

gpx1000 committed:
Merge branch 'master-upstream'

gpx1000 committed:
Merge branch 'master-upstream'

gpx1000 committed:
Merge branch 'master-upstream'

gpx1000 committed:
run clang-format

gpx1000 committed:
fix android

gpx1000 committed:
fix android

gpx1000 committed:
Merge branch 'master-upstream'

gpx1000 committed:
address the returns doc.

gpx1000 committed:
Merge branch 'KhronosGroup:master' into master

gpx1000 committed:
Merge branch 'upstream-master'

gpx1000 committed:
Merge branch 'KhronosGroup:master' into master

gpx1000 committed:
Merge branch 'upstream-master'

gpx1000 committed:
Merge branch 'upstream-master'

gpx1000 committed:
Merge branch 'KhronosGroup:master' into master

gpx1000 committed:
Merge remote-tracking branch 'origin/master'

gpx1000 committed:
Fix issue #452

gpx1000 committed:
Merge branch 'upstream-master'

JeremyG2101 committed:
PLEASE DO NOT MERGE IT! This is a Debug sample including: memory_budget

JeremyG2101 committed:
working version update

JeremyG2101 committed:
Quick fix just removed 2 test log lines

JeremyG2101 committed:
All changes were made based on Steve's comments. Syntax warnings were claned from my end. All printf related functions/commands were removed.

JeremyG2101 committed:
copyrights year should be 2019 based on the instancing created

JeremyG2101 committed:
CMake list update.

JeremyG2101 committed:
add copyrights to CMake List

JeremyG2101 committed:
Per Steve's Comments: 1) In the CMakeList, the introduction is now 100% focus on what was introduced related to the memory budget extension feature. 2)update_memory_properties() function is now REMOVED from the draw() and reintroduced to the prepare_instance_data() function, after everything is calculated, loaded, and staged. 3) All runtime update/button-press-update related boolean and functions were REMOVED.

JeremyG2101 committed:
The Tutorial is fianlly done, and the image is saved in .PNG format, organized in an individual folder since the memory_budget folder. the tutorial is named as README.md

JeremyG2101 committed:
This update regards to Steve's newest comments to revise the description modification copyrights to Holochip Corporation; and insert additional description of the usage of update_device_memory_properties().

JeremyG2101 committed:
Quick syntax fix on "the" to "The" from description of int variable "mesh_density".

JeremyG2101 committed:
went thfough the clang check and already have the format modified, our project name is added to the CmakeList

JeremyG2101 committed:
Merge pull request #8 in holochip/vulkan-samples from Memory_Budget_Extension to master

Memory_Budget_Debug_Sample

Approved by Steven Winston

JeremyG2101 committed:
FR-406: Memory Budget: Quick fix corresponding for the GitHub comments regarding memory units conversion issues and get_memory_properties2 call inside prepare()

JeremyG2101 committed:
Code revised per Steve's new comments 10/25/2022

JeremyG2101 committed:
This is the dummy renderer class built for full_screen_exclusive extension sample, after it passed check, I will create an individual branch just for a dummy renderer, it I don't have to write 1,000+ lines to begin with. Where I can directly dive into add models and change skybox.

JeremyG2101 committed:
So this is the clean and neat version of it, Clang check applied.

JeremyG2101 committed:
code revised and set private to extension related variables

JeremyG2101 committed:
Full Screen Extension exclusive

JeremyG2101 committed:
Organized the full Screen Exclusive branch, which it is actually corresponding to the current/up-to-date changes of the memory_budget. It also adds the full_screen_exclusive to the project CMake list.

JeremyG2101 committed:
Adds more annotations for display mode selection

JeremyG2101 committed:
noticed an incorrect annotation

gpx1000 committed:
Merge branch 'KhronosGroup:master' into master

JeremyG2101 committed:
A clean version of this

gpx1000 committed:
Merged master into Full_Screen_Exclusive_Extension

gpx1000 committed:
Merge branch 'upstream-master'

gpx1000 committed:
Merge remote-tracking branch 'origin/master'

gpx1000 committed:
Merge branch 'master' into downstream-master

# Conflicts:
#	samples/CMakeLists.txt

gpx1000 committed:
Revert memory budget should NOT be merged to master.

gpx1000 committed:
Revert "This update regards to Steve's newest comments to revise the description modification copyrights to Holochip Corporation; and insert additional description of the usage of update_device_memory_properties()."

This reverts commit 889ece87b111e808e71a4d98069fe7df1ccc0b68.

gpx1000 committed:
Revert "Quick syntax fix on "the" to "The" from description of int variable "mesh_density"."

This reverts commit 35e7dae9ff8a859a1905867236dd88fc6cd22fd2.

gpx1000 committed:
Revert "The Tutorial is fianlly done, and the image is saved in .PNG format, organized in an individual folder since the memory_budget folder. the tutorial is named as README.md"

This reverts commit 3cc5023045cf19ee3ad1e011f35ae84b394d9597.

gpx1000 committed:
Revert "Per Steve's Comments: 1) In the CMakeList, the introduction is now 100% focus on what was introduced related to the memory budget extension feature. 2)update_memory_properties() function is now REMOVED from the draw() and reintroduced to the prepare_instance_data() function, after everything is calculated, loaded, and staged. 3) All runtime update/button-press-update related boolean and functions were REMOVED."

This reverts commit 8cdc982c22e040ace07e62a866758e3e58e99a81.

gpx1000 committed:
Revert "copyrights year should be 2019 based on the instancing created"

This reverts commit 428cc6eab40f504a1b846995f1e9274051daae62.

gpx1000 committed:
Revert "add copyrights to CMake List"

This reverts commit 1086cab586b2dfa78eac2b111c5ff4d39779fec9.

gpx1000 committed:
Revert "CMake list update."

This reverts commit 8c14c62074c1bb713e35da3f546a6a5009de9a90.

gpx1000 committed:
Revert "All changes were made based on Steve's comments. Syntax warnings were claned from my end. All printf related functions/commands were removed."

This reverts commit 8970f8fdce9eaa76d743748982be4d06e6a5be61.

gpx1000 committed:
Revert "Quick fix just removed 2 test log lines"

This reverts commit 7708fff86dd2adf80aa37efb9e122b117c18cd7d.

gpx1000 committed:
Revert "working version update"

This reverts commit 9a56f3b7bb387f683d6b08bb89becb1110e7256d.

gpx1000 committed:
Revert "PLEASE DO NOT MERGE IT! This is a Debug sample including: memory_budget"

This reverts commit f6d3730ebcfe9a4f96451133e18fddf8b3af03ef.

gpx1000 committed:
Merge branch 'downstream-master' into Full_Screen_Exclusive_Extension

JeremyG2101 committed:
added all fixes according to the last commonts. This is also a quick save point before I add the customized swapchain, swapchain image, and image view functions. The mentioned function set will ultimately replaces the swapchain and image views created from the render_context class, but remains everything else the same.

JeremyG2101 committed:
full screen exclusive are updated based on framework changes made in the most up-to-date vulkan sample version. In addition to which, 3 bugs are fixed (suggested to fix) in response to the newest version of glslang, regarding to the directory/directories change(s) made to LimitedResources.

JeremyG2101 committed:
This update adds an entire image view recreate function in addition to the swapchain recreation. Meantime, the swapchain images are are deep copied to a new image and sync them after the swapchain recreation. Clang checked for both files.

JeremyG2101 committed:
Add a boolean to activate or deactive swapchain and full screen exclusive related functions

JeremyG2101 committed:
Calibrated Time Stamps

JeremyG2101 committed:
Introduces update_time_domain, read_time_domain, and update_timestamps. Several vectors and structures are introduced as well. All syntax organized in Clang format.

JeremyG2101 committed:
Quick Save: add some veriations to the subpass, added and modified calibrated timestamps related functions, device extension debug, sample CMake list updated

JeremyG2101 committed:
quick save adding some modifications

JeremyG2101 committed:
Quick Save: this pretty much finalizes the entire sample, its tutorial will be updated shortly.

JeremyG2101 committed:
Adds README.md file, and its reference image; minior formating issues and Clang format checked.

JeremyG2101 committed:
fixed some wording issues from README.md

JeremyG2101 committed:
Vulkan Sample framework update corresponding to the latest Volk and Glslang version.

JeremyG2101 committed:
This is a raw renderer build for enableing mesh shader. rendering pass spot saved for introducing mesh shader, and many more are reserved for this feature to be added.

JeremyG2101 committed:
Quick save, made some changes on the cpp.

JeremyG2101 committed:
Added 1) task compile shader, 2) draw cull compile shader, 3) task shader, 4) mesh shader, 5) frag/pixel shader, 6) utils which includes structures and functions that share usage across most of shaders, that are mesh_shader related.

JeremyG2101 committed:
Quick fix: some syntax issue from mesh task shaders

JeremyG2101 committed:
Added mesh shader pipeline and pending on setting off an extra pipeline layout, Task shader header and cull shader are still needed!

JeremyG2101 committed:
Add some more functionalities for compute pipeline, and added a set of TODO list for command buffer

JeremyG2101 committed:
Quick Save: traditional N body for loop staging with traditional graphic pipeline, using vertex and index created insisde the class.

JeremyG2101 committed:
Quick Save: added init_cube function, added prepare_cube_buffers function

JeremyG2101 committed:
Quick Save: 1) Meshlet structure and veriables are defined 2) init_cube_meshlet function introduced

JeremyG2101 committed:
a wording issue fixed

JeremyG2101 committed:
remove changes from framework and its related files.

JeremyG2101 committed:
remove framework modifications

JeremyG2101 committed:
Quick Save: this is a pre-finalizing stage, basically added the mesh shaders capablities.

JeremyG2101 committed:
a syntax and format change

JeremyG2101 committed:
Quick save: this marks out several whereabouts to correct the object buffers/descriptor sets/layers/layer bindings, graphic pipeline, and command buffer.

JeremyG2101 committed:
Quick Save: Fixed everything renderer and Cpp related, and is bug free

JeremyG2101 committed:
Quick Save

JeremyG2101 committed:
Quick save

JeremyG2101 committed:
Quick Save

JeremyG2101 committed:
add some syntax fix on shaders

JeremyG2101 committed:
syntax update

JeremyG2101 committed:
quick save

JeremyG2101 committed:
quick save

JeremyG2101 committed:
Quick Save

JeremyG2101 committed:
Quick Save: renderer built, uniform buffer built, gui built, kepboard movement control built. runtime: pass

JeremyG2101 committed:
Quick Save: .mesh and .task modified for the meshlet and ubo concept.

JeremyG2101 committed:
quick save: added the mesh shader generation for a emit id 1,1,1

JeremyG2101 committed:
quick save for shader updates

JeremyG2101 committed:
This is the finalized version of the sample code.

JeremyG2101 committed:
Finalizes the README.md, and some syntax issues were fixed.

JeremyG2101 committed:
some syntax changes and Clang check

JeremyG2101 committed:
I accidently mod an unrelated file....

gpx1000 committed:
quick pass at word smithing the tutorial.

Also updated the CMakeLists.txt

Mesh shading only requires Vulkan 1.1 not 1.3 corrections made.

gpx1000 committed:
Merge branch 'main' into vk_Mesh_Shader_Culling

# Conflicts:
#	samples/CMakeLists.txt

gpx1000 committed:
add in information and resources.

gpx1000 committed:
revert submodules to main

gpx1000 commented:
This PR depends upon #541 being merged first.  It will not build correctly until changes in there are merged into here.

gpx1000 committed:
glslang had some breaking changes with the methodology we were using of grabbing the resource limits.  Updating to the latest glslang and resolving the breaking changes as per glslang recommendations.
This PR now supersedes the changes in the #541

gpx1000 committed:
fix copyright

gpx1000 committed:
fix clang-format

gpx1000 committed:
volk needs to be updated.

gpx1000 committed:
more clang-format fixes.

gpx1000 commented:
glslang was updated with a breaking change for us.  Given that #541 has one approval, I figured I'd do the required changes here and get glslang to the latest version. Thus we need to merge #541 and then this to capture the latest glslang required updates.

This PR is ready to review on its' own now.

SaschaWillems commented on samples/extensions/mesh_shader_culling/README.md:
````diff
@@ -0,0 +1,221 @@
+
+## Overview
+
+This sample demonstrates how to incorporate the Vulkan extension ```VK_EXT_mesh_shader_exclusive ```, and
````
Should read `VK_EXT_mesh_shader`.

SaschaWillems commented on samples/extensions/mesh_shader_culling/README.md:
````diff
@@ -0,0 +1,221 @@
+1) ```VK_KHR_GET_PHYSICAL_DEVICE_PROPERTIES_2```
+2) ```VK_KHR_SPIRV_1_4```
+3) ```VK_EXT_MESH_SHADER```
+4) ```VK_API_VERSION_1_1```
````
Strictly spoken this is not an extension, maybe move it out of this list and simply list "Vulkan 1.1" as a base requirement.

SaschaWillems commented on samples/extensions/mesh_shader_culling/README.md:
```diff
@@ -0,0 +1,221 @@
+## Task Shader
+
+A simple task shader processes the data from the uniform 
+buffer, and then send it to its associated mesh
```
sends instead of send

SaschaWillems commented on samples/extensions/mesh_shader_culling/README.md:
```diff
@@ -0,0 +1,221 @@
+
+* A task shader is used to emit mesh shader(s) (or not)
+* A mesh shader has the responsibility to generate 
+  primatives.
```
primitives instead of primatives

SaschaWillems commented on samples/extensions/mesh_shader_culling/README.md:
```diff
@@ -0,0 +1,221 @@
+// the actual position of each meshlet:
+vec4 position = displacement + sharedData.positionTransformation;
+float squareRadius = position.x * position.x + position.y * position.y;
+// Cull Logic: only is the meshlet center position is within the view circle defined by the cull radius,
```
Only if instead of only is

SaschaWillems commented on shaders/mesh_shader_culling/mesh_shader_culling.mesh:
```diff
@@ -0,0 +1,115 @@
+        vec4 position = displacement + sharedData.positionTransformation;
+        float squareRadius = position.x * position.x + position.y * position.y;
+
+        // Cull Logic: only is the meshlet center position is within the view circle defined by the cull radius,
```
Only if instead of only is

SaschaWillems commented on samples/extensions/mesh_shader_culling/README.md:
````diff
@@ -0,0 +1,221 @@
+A simple UI overlay is created to help showcase the mesh shader cull feature. Where, the radius of the circular cull
+mask can be adjusted using the "Cull Radius" slider, which ranges from ```0.5f``` to ```2.0f```. And the meshlet density
+level can be controlled using the combo-box. Where, a selection of 4-by-4, 6-by-6, and 8-by-8 of density opinions is
+provided. Where, users may use keyboard input ```W``` ```A``` ```S``` ```D``` to move the square mesh around the scene.
````
Maybe it's because I'm not a native speaker, but those "Where," make this cumbersome to read.

SaschaWillems commented on samples/extensions/mesh_shader_culling/mesh_shader_culling.cpp:
```diff
@@ -0,0 +1,317 @@
+
+void MeshShaderCulling::setup_descriptor_pool()
+{
+	// I still build the pool using a vector, since practically it cannot be just one descriptor pool size.
```
This comment (starting with "I still build..." is used multiple times. It sounds like a left-over.

SaschaWillems commented on samples/extensions/mesh_shader_culling/README.md:
````diff
@@ -0,0 +1,221 @@
+Thus, we disable the ```pVertexInputState``` and 
+```pInputAssemblyState```.
+
+It is expected the normal use case would be to send 
````
What exactly does this mean?

SaschaWillems reviewed (changes requested):
The sample itself works fine on my system. I added a few comments though, mostly for the readme and code comments. In general the readme feels a bit cumbersome to read, at least for me as a non-native speaker.

SaschaWillems commented on samples/extensions/mesh_shader_culling/mesh_shader_culling.cpp:
```diff
@@ -0,0 +1,317 @@
+	// Adding device extensions
+	add_device_extension(VK_KHR_SPIRV_1_4_EXTENSION_NAME);
+	add_device_extension(VK_EXT_MESH_SHADER_EXTENSION_NAME);
+	add_device_extension(VK_KHR_SHADER_FLOAT_CONTROLS_EXTENSION_NAME);
```
Is this really required for this sample?

SaschaWillems commented:
The sample also need to be added to the samples listing readme.

gpx1000 committed:
update glfw

gpx1000 committed:
rework the tutorial.  Add the blurb to the main samples README and fix some typos in the cpp.

gpx1000 commented:
@SaschaWillems If you get a chance, I just reworked the tutorial.  Please give it a read and let me know if that is easier to understand.  All comments should be addressed from above as well.

gpx1000 committed:
Merge branch 'main' into vk_Mesh_Shader_Culling

SaschaWillems commented:
TVM. This reads a lot better now! 👍

asuessenbach commented:
I would like to review this, but when I hit the "Start a review" button or the "Add single comment" button in the "Files changed" tab nothing happens.
Any ideas?

gpx1000 commented:
I'm not sure, it seems to work okay when I try to do it here.  If you have clion or Android studio you can do reviews through the GitHub plugin.

marty-johnson59 commented:
Hi @aurolou, can you test this one out on AMD?  This is a DT sample - would be good to make sure it's OK on your HW.  Also, please look at Sascha's descriptor buffer sample again (PR 600) - hopefully the TDR issues is fixe now.  Thanks for your help!

asuessenbach commented:
I tried to configure a solution for this PR (Win10, VS2022, using CMake GUI), but it fails. It seems to miss pthread.h and pthreads.lib.
Any advice how to resolve that?

SaschaWillems commented:
@asuessenbach: Did you do a clean checkout and cmake cache delete? I've been having similar troubles with other PRs in the past cause I'm usually just switching PR branches.

asuessenbach commented:
@SaschaWillems Indeed, deleting the caches and updating some submodules did help. Thanks for that hint.

asuessenbach commented on samples/extensions/mesh_shader_culling/mesh_shader_culling.cpp:
```diff
@@ -0,0 +1,310 @@
+	set_api_version(VK_API_VERSION_1_1);
+
+	// Adding instance extension
+	add_instance_extension(VK_KHR_GET_PHYSICAL_DEVICE_PROPERTIES_2_EXTENSION_NAME);
```
No need to add `VK_KHR_GET_PHYSICAL_DEVICE_PROPERTIES_2_EXTENSION_NAME` with `VK_API_VERSION_1_1`

asuessenbach commented on samples/extensions/mesh_shader_culling/mesh_shader_culling.cpp:
```diff
@@ -0,0 +1,310 @@
+	// Adding instance extension
+	add_instance_extension(VK_KHR_GET_PHYSICAL_DEVICE_PROPERTIES_2_EXTENSION_NAME);
+	// Adding device extensions
+	add_device_extension(VK_KHR_SPIRV_1_4_EXTENSION_NAME);
```
using `VK_KHR_SPIRV_1_4_EXTENSION_NAME` also requires `VK_KHR_SHADER_FLOAT_CONTROLS_EXTENSION_NAME`

asuessenbach commented on framework/glslang_default_resource_limits.h:
This file should be added to the framework/CMakeLists.txt.
And: why do you introduce it at all and don't use `GetDefaultResources()` from `third_party\glslang\glslang\Public\ResourceLimits.h` ?

gpx1000 replied:
technically headerfiles shouldn't need to be added to targets in CMake.  CMake is designed to find those header files as needed.  However, thanks for the suggestion to use GetDefaultResources.  I've removed the new header file and used that throughout.

asuessenbach commented on samples/CMakeLists.txt:
```diff
@@ -48,15 +48,15 @@ endforeach ()
+        "instancing"
+        "compute_nbody"
+        "terrain_tessellation"
+        "hlsl"
```
No need to indent those samples.

asuessenbach commented on samples/CMakeLists.txt:
```diff
@@ -77,30 +78,30 @@ set(ORDER_LIST
-    #Tooling samples
-    "profiles"
+        #Tooling samples
+        "profiles"
```
Indenting again.

asuessenbach commented on samples/api/hpp_hlsl_shaders/hpp_hlsl_shaders.h:
```diff
@@ -1,4 +1,4 @@
-/* Copyright (c) 2022, NVIDIA CORPORATION. All rights reserved.
+/* Copyright (c) 2022-2023, NVIDIA CORPORATION. All rights reserved.
```
No need to just adjust the copyright, when nothing else changed in this file.

asuessenbach commented on samples/extensions/mesh_shader_culling/README.md:
````diff
@@ -0,0 +1,188 @@
+To enable the mesh shading feature, the following 
+extensions are required: (NB: ```VK_API_VERSION_1_1``` is required as a base requirement)
+
+1) ```VK_KHR_GET_PHYSICAL_DEVICE_PROPERTIES_2```
````
With `VK_API_VERSION_1_1`, you don't need `VK_KHR_GET_PHYSICAL_DEVICE_PROPERTIES_2_EXTENSION_NAME` any more, as it's promoted into that version

asuessenbach commented on samples/extensions/mesh_shader_culling/README.md:
````diff
@@ -0,0 +1,188 @@
+extensions are required: (NB: ```VK_API_VERSION_1_1``` is required as a base requirement)
+
+1) ```VK_KHR_GET_PHYSICAL_DEVICE_PROPERTIES_2```
+2) ```VK_KHR_SPIRV_1_4```
````
`VK_KHR_SPIRV_1_4_EXTENSION_NAME` also requires `VK_KHR_SHADER_FLOAT_CONTROLS_EXTENSION_NAME`

asuessenbach commented on samples/extensions/mesh_shader_culling/README.md:
````diff
@@ -0,0 +1,188 @@
+
+1) ```VK_KHR_GET_PHYSICAL_DEVICE_PROPERTIES_2```
+2) ```VK_KHR_SPIRV_1_4```
+3) ```VK_EXT_MESH_SHADER```
````
To list the extension here, you should either use their actual name (`VK_KHR_spirv_1_4`) or their named define (`VK_KHR_SPIRV_1_4_EXTENSION_NAME`)

asuessenbach commented on samples/extensions/mesh_shader_culling/README.md:
```diff
@@ -0,0 +1,188 @@
+    float subDimension;
+    float cullRadius;
+};
+// 2) use the following command to "establish the connection" to its associated mesh shader:
```
It's not exactly a command. It's more like a variable with a storage class specifier

asuessenbach commented on samples/extensions/mesh_shader_culling/README.md:
```diff
@@ -0,0 +1,188 @@
+
+More details about emitting a mesh task can be found in the attached article:
+
+[Introduction to Turing Mesh Shaders](https://developer.nvidia.com/blog/introduction-turing-mesh-shaders/)
```
Maybe also mention this article: [Mesh Shading for Vulkan](https://www.khronos.org/blog/mesh-shading-for-vulkan)

asuessenbach commented on samples/extensions/mesh_shader_culling/README.md:
```diff
@@ -0,0 +1,188 @@
+* [Meshlet Size tradeoffs](https://zeux.io/2023/01/16/meshlet-size-tradeoffs/)
+* NVIDIA - [Mesh Shaders in Turing](https://on-demand.gputechconf.com/gtc-eu/2018/pdf/e8515-mesh-shaders-in-turing.pdf)
+* AMD - [Sampler feedback ultimate in Mesh shaders](https://gpuopen.com/wp-content/uploads/slides/AMD_RDNA2_DirectX12_Ultimate_SamplerFeedbackMeshShaders.pdf)
+* [Timur's blog](https://timur.hu/blog/2022/mesh-and-task-shaders)
```
We're probably allowed to link to AMD and NVIDIA pages. But what about the other two?

gpx1000 replied:
That's actually a good question.  I know Timur was commenting on this so @Venemo do you mind if we refernce your blog article?  I'll remove the other reference.

Venemo replied:
I think it's fine to mention my article. Let's ask @zeux about the other one.

zeux replied:
Feel free to link to my article.

gpx1000 replied:
Thanks guys, much appreciated.  I'll leave both of your articles in.

asuessenbach commented on samples/extensions/mesh_shader_culling/README.md:
```diff
@@ -0,0 +1,188 @@
+
+## Task Shader
+
+A task shader is an optional stage responsible for launching mesh shaders.  It has two purposes:
```
As you mention, the task shader is optional. Maybe you should also mention what's done when you don't provide one.

Venemo replied:
Additionally, please mention that task shaders are highly recommended in order to get the best performance out of a mesh shading pipeline.

Venemo replied:
I like the additional paragraph. I also recommend to modify the above sentence to: "A task shader is an optional but recommended stage...." to emphasize this even better.

asuessenbach commented on samples/extensions/mesh_shader_culling/README.md:
```diff
@@ -0,0 +1,188 @@
+-->
+
+# Mesh Shader Culling
+
```
Maybe you could add an image of this sample here, to show the expected results.

asuessenbach replied:
You need to also add the image to the PR.

asuessenbach commented on samples/extensions/mesh_shader_culling/mesh_shader_culling.cpp:
```diff
@@ -0,0 +1,310 @@
+	if (camera.keys.left || camera.keys.right || camera.keys.up || camera.keys.down)
+	{
+		ubo_cull.cull_center_x = (-1.0f) * camera.position.x;
+		ubo_cull.cull_center_y = (-1.0f) * camera.position.z;
```
Maybe a little comment: why is `cull_center` the negative `camera.position`?
And: why do you explicitly multiply with `(-1.0f)`, instead of just `-camera.position.x`

asuessenbach commented on samples/extensions/mesh_shader_culling/mesh_shader_culling.h:
```diff
@@ -0,0 +1,65 @@
+
+class MeshShaderCulling : public ApiVulkanSample
+{
+  private:
```
Why is just the `density_level` private and all the other data public?
I would have expected every data to be private.

asuessenbach commented on shaders/mesh_shader_culling/mesh_shader_culling.mesh:
```diff
@@ -0,0 +1,115 @@
+    int   meshletsNumber;
+    float subDimension;
+    float cullRadius;
+};
```
Maybe move the `struct SharedData` into a header and include that in both, the task and the fragment shader?

asuessenbach commented on shaders/mesh_shader_culling/mesh_shader_culling.mesh:
```diff
@@ -0,0 +1,115 @@
+// 2) index data associated with the vertex data, triangle defined using right-hand rule
+
+const uvec3 unitPrimitive_0 = uvec3(0, 1, 2);
+const uvec3 unitPrimitive_1 = uvec3(2, 0, 3);
```
The two triangles have different winding. Doesn't matter here, but it would be better to use `(2, 3 0)` for the second triangle.

asuessenbach commented on shaders/mesh_shader_culling/mesh_shader_culling.task:
```diff
@@ -0,0 +1,96 @@
+
+#extension GL_EXT_mesh_shader: require
+
+layout(local_size_x = 64, local_size_y = 1, local_size_z = 1) in;
```
Why are you using `local_size_x = 64`, while `vkCmdDrawMeshTasksEXT` is called with `num_workgroups_x = 1`?

asuessenbach commented on shaders/mesh_shader_culling/mesh_shader_culling.task:
```diff
@@ -0,0 +1,96 @@
+
+    if (ubo.meshlet_density == 0)
+    {
+        // Attempt to create a 6 by 6 meshlets array
```
4 by 4 ?

asuessenbach commented on shaders/mesh_shader_culling/mesh_shader_culling.task:
```diff
@@ -0,0 +1,96 @@
+    }
+    else if (ubo.meshlet_density == 2)
+    {
+        // Attempt to create a 6 by 6 meshlets array
```
8 by 8?

asuessenbach commented on shaders/mesh_shader_culling/mesh_shader_culling.mesh:
```diff
@@ -0,0 +1,115 @@
+
+#extension GL_EXT_mesh_shader: require
+
+layout(local_size_x = 64, local_size_y = 1, local_size_z = 1) in;
```
local_size_x = 1 ?

asuessenbach commented on shaders/mesh_shader_culling/mesh_shader_culling.mesh:
```diff
@@ -0,0 +1,115 @@
+#extension GL_EXT_mesh_shader: require
+
+layout(local_size_x = 64, local_size_y = 1, local_size_z = 1) in;
+layout(triangles, max_vertices = 64 * 4, max_primitives = 124 * 2) out;
```
max_primitives = 64 * 2 ?

asuessenbach reviewed (commented):
Overall, this is a good example for mesh culling.
Currently, you spawn one mesh shader per task shader (with just one task shader). Maybe the value of mesh shaders would be better represented by spawning N*N mesh shaders from your task shader?

marty-johnson59 commented:
Hi @aurolou, could you check this sample on AMD HW before we merge?  We think it's ready.   Also, if you could re-run the descriptor buffer sample ( [PR 600: ](https://github.com/KhronosGroup/Vulkan-Samples/pull/600) ), that one is hopefuly working now too.  Thanks!

cforfang commented on shaders/mesh_shader_culling/mesh_shader_culling.mesh:
```diff
@@ -0,0 +1,115 @@
+        // Cull Logic: only is the meshlet center position is within the view circle defined by the cull radius,
+        // then this very meshlet will be generated. Otherwise, meshlet will NOT be generated.
+
+        if (squareRadius < sharedData.cullRadius * sharedData.cullRadius)
```
If I'm reading this right then some elements of `gl_MeshVerticesEXT[]` and `gl_PrimitiveTriangleIndicesEXT[]` are not written to depending on this culling logic, is that right? However, `SetMeshOutputsEXT()` is always the same value regardless of the culling logic?

Is the implication then that the consumer of `gl_MeshVerticesEXT` and `gl_PrimitiveTriangleIndicesEXT` may be reading either (a) some undefined values (if the culling logic caused those elements to _not_ be written to), or (b) defined-to-be-something (zero?) values? 

Basically, I'm wondering if this is well-defined (specifically, if `gl_MeshVerticesEXT` and `gl_PrimitiveTriangleIndicesEXT` have well-defined default values and if this is relied on here) -- and also if this could really be called _culling_ when it seems more like the same number of primitives are always output, it's just than when 'culled' they are set up as degenerate primitives (thus _processed_ but not actually visible)?

asuessenbach commented on samples/extensions/mesh_shader_culling/mesh_shader_culling.cpp:
```diff
@@ -0,0 +1,310 @@
+		{
+			ubo_cull.meshlet_density = static_cast<float>(density_level);
+			update_uniform_buffers();
+		}
```
Maybe you could add some pipeline statistics here (FS, TS, MS invocations, for example)

asuessenbach reviewed (commented):
The current approach with one mesh shader invocation performing a loop does not really reflect the power of task/mesh shaders. Maybe you could change it to have N*N mesh shaders.
Besides that, some pipeline statistics like number of Task Shader, Mesh Shader, and Fragment Shader invocations could be of interest.
If you like, I could send you a patch with a few modifications to showcase that.
Finally, I think it might be worth to substantially increase the fragment shader complexity, to actually get some improvement with the mesh shader approach.

asuessenbach commented on samples/extensions/mesh_shader_culling/mesh_shader_culling.cpp:
```diff
@@ -0,0 +1,310 @@
+}
+
+void MeshShaderCulling::request_gpu_features(vkb::PhysicalDevice &gpu)
+{
```
You should check for taskShader and meshShader support here.

Venemo commented on samples/extensions/mesh_shader_culling/README.md:
````diff
@@ -0,0 +1,188 @@
+when working with models.
+
+Thus, we disable the ```pVertexInputState``` and 
+```pInputAssemblyState```.
````
Suggestion:

Thus, we disable the ```pVertexInputState``` and 
```pInputAssemblyState``` by setting them to NULL.

Venemo commented on samples/extensions/mesh_shader_culling/README.md:
```diff
@@ -0,0 +1,188 @@
+A task shader is an optional stage responsible for launching mesh shaders.  It has two purposes:
+
+* Decide how many mesh shaders to launch in the workgroup.
+* Create task payload to share with the mesh shader.
```
I think "share" does not describe well what is happening. Suggestion:

"Create a task payload that mesh shaders will have read-only access to."

Venemo commented on samples/extensions/mesh_shader_culling/README.md:
```diff
@@ -0,0 +1,188 @@
+* Decide how many mesh shaders to launch in the workgroup.
+* Create task payload to share with the mesh shader.
+  * taskPayloadSharedEXT type variable in GLSL can exist at most once it gets implicitly used by EmitMeshTasksEXT 
+    and behaves like shared memory.
```
Suggestion:

... behaves like shared memory in the task shader and like a read-only SSBO in a mesh shader.

Venemo commented on samples/extensions/mesh_shader_culling/README.md:
```diff
@@ -0,0 +1,188 @@
+with an (optional) task shading stage that can call other mesh 
+shading stages.
+
+* A task shader (optional) is used to emit mesh shader(s)
```
Suggestion to clarify this sentence:

A task shader (optional) is used to launch mesh shader workgroup(s)

Venemo commented on samples/extensions/mesh_shader_culling/README.md:
```diff
@@ -0,0 +1,188 @@
+
+* A task shader (optional) is used to emit mesh shader(s)
+* A mesh shader has the responsibility to generate 
+  primitives.
```
and vertices

Venemo commented on samples/extensions/mesh_shader_culling/README.md:
```diff
@@ -0,0 +1,188 @@
+    int   meshletsNumber;
+    float subDimension;
+    float cullRadius;
+};
```
It should be mentioned that all GPU vendors recommend to use as little task payload as possible, eg. by packing the data to fewer bits etc.

Venemo commented on samples/extensions/mesh_shader_culling/README.md:
```diff
@@ -0,0 +1,188 @@
+
+## Mesh Shader
+
+A mesh shader executes on the same workgroup as the task shader that emitted it if a task shader exists. A 
```
This is not true. The mesh shader is NOT part of the same workgroup as the task shader. One task shader workgroup can launch many workgroups of mesh shaders (which can read the payload from the task workgroup that launched them).

Venemo commented on samples/extensions/mesh_shader_culling/README.md:
```diff
@@ -0,0 +1,188 @@
+
+A mesh shader executes on the same workgroup as the task shader that emitted it if a task shader exists. A 
+mesh shader is responsible for generating 
+vertices and indices 
```
vertices and primitives (primitives are represented by their vertex indices)

Venemo commented on samples/extensions/mesh_shader_culling/README.md:
```diff
@@ -0,0 +1,188 @@
+
+[Using Mesh Shaders for Professional Graphics](https://developer.nvidia.com/blog/using-mesh-shaders-for-professional-graphics/)
+
+## Meshlets culling
```
What is described below is not meshlet culling but per-primitive culling.

To clarify:

* Per-meshlet culling is done in the task shader. One example of how this can be done is having a bounding box of the entire meshlet and test whether the entire meshlet would be invisible. This is highly recommended because it reduces the number of mesh shader workgroups launched by that shader, so it saves compute (and bandwidth) resources by not launching any mesh shader workgroups for meshlets that would be invisible.
* Per-primitive culling is done in the mesh shader and its purpose is to have the mesh shader emit fewer vertices and primitives. This is typicaly less beneficial and is only recommended for scenes with very high geometric complexity or on GPUs whose fixed-function culling is slower than shaders. This is no longer recommended by neither AMD nor NVidia for their latest generation GPUs.

gpx1000 commented:
> 

Sure,  I always accept help when it's offered :)

gpx1000 committed:
address all comments.
Special thanks to @asuessenbach for submitting a patch.

Add capability to use include in shaders.
Add shared header for the mesh and task shader to share.
Use getDefaultResources instead of a new header file.

gpx1000 committed:
clang format fix.

Venemo commented on samples/extensions/mesh_shader_culling/README.md:
````diff
@@ -0,0 +1,198 @@
+## Overview
+
+This sample demonstrates how to incorporate the Vulkan extension [```VK_EXT_mesh_shader```](https://www.khronos.org/registry/vulkan/specs/1.2-extensions/html/vkspec.html#VK_EXT_mesh_shader), and
+introduces mesh shader culling using meshlets.
````
The term "mesh shader culling" is vague, it should either be:

* per-meshlet culling in a task shader, or
* per-primitive culling in a mesh shader

Venemo replied:
(or both)

Venemo commented on samples/extensions/mesh_shader_culling/README.md:
```diff
@@ -0,0 +1,198 @@
+## Mesh Shader
+
+A mesh shader does not execute on the same workgroup as the task shader that emitted it if a task shader exists.  
+Many workgroups can be  One task shader workgroup can launch many mesh shader workgroups.
```
I think this paragraph still needs editing. Suggestion:

Task and mesh shaders are executed in workgroups similar to compute shaders. Each task shader workgroup can launch many mesh shader workgroups. Each mesh shader workgroup is responsible for generating vertices and primitives. The API allows for any logic, but a typical application it is recommended to organize this around meshlets, which are a small group of vertices and primitives. Typically, each task shader invocation processes a group of meshlets and each mesh shader workgroup processes one meshlet.

Venemo commented on samples/extensions/mesh_shader_culling/README.md:
```diff
@@ -0,0 +1,198 @@
+method of culling is to decide via a condition if a meshlet should be generated or not.
+
+In this sample, a circular visual zone is centered at the origin, with an adjustable radius, controlled by the gui.
+When a meshlet moves out of the visual zone, its generation process will be skipped.
```
Sorry but this is still unclear to me from this text. Does the sample implement per-primitive or per-meshlet culling? Or maybe both?

asuessenbach commented on samples/extensions/mesh_shader_culling/README.md:
````diff
@@ -0,0 +1,198 @@
+extensions are required: (NB: ```VK_API_VERSION_1_1``` is required as a base requirement)
+
+1) ```VK_KHR_SPIRV_1_4_EXTENSION_NAME```
+2) ```VK_EXT_MESH_SHADER```
````
`VK_EXT_MESH_SHAEDER` -> `VK_EXT_MESH_SHADER_EXTENSION_NAME`

asuessenbach commented on samples/extensions/mesh_shader_culling/mesh_shader_culling.cpp:
```diff
@@ -0,0 +1,429 @@
+
+void MeshShaderCulling::request_gpu_features(vkb::PhysicalDevice &gpu)
+{
+	// Query whether the device supports buffer device addresses
```
"Query" -> "Check"
"buffer device addresses" -> "task and mesh shaders"

gpx1000 committed:
Revise the readme per the requested changes.

Venemo reviewed (commented):
The readme looks good to me now.

asuessenbach commented on shaders/mesh_shader_culling/mesh_shader_culling.mesh:
```diff
@@ -0,0 +1,94 @@
+
+#include "mesh_shader_culling/mesh_shader_shared.h"
+
+// 2) using the following command to "establish the connection" (variable name can vary):
```
`2)` ?

asuessenbach commented on shaders/mesh_shader_culling/mesh_shader_culling.task:
```diff
@@ -0,0 +1,84 @@
+} ubo;
+
+
+// 2) use the following command to "establish the connection" to its associated mesh shader:
```
`2)` ?

asuessenbach commented on shaders/mesh_shader_culling/mesh_shader_shared.h:
```diff
@@ -0,0 +1,28 @@
+	int   meshletsNumber;
+	float subDimension;
+	float cullRadius;
+};
```
In order to reduce payload size: `N` and `meshletsNumber` are not used in the mesh shader, `subDimension` could be easily calculated out of a constant and `gl_WorkGroupSize`. And from `positionTransformation`, only `x` and `y` need to be carried. Thus the size of `SharedData` could be reduced from 8 * 4 byte to 3 * 4 byte.
And one could even argue, that those three values are part of an UBO, that could be passed into the mesh shader as well, instead of the payload. But in order demonstrate how a payload is managed, it should probably be kept, maybe with a comment on how to remove it.

asuessenbach commented on samples/extensions/mesh_shader_culling/README.md:
````diff
@@ -23,7 +23,7 @@
 
 This sample demonstrates how to incorporate the Vulkan extension [```VK_EXT_mesh_shader```](https://www.khronos.org/registry/vulkan/specs/1.2-extensions/html/vkspec.html#VK_EXT_mesh_shader), and
-introduces mesh shader culling using meshlets.
+introduces per primative culling in a mesh shader.
````
primative -> primitive

gpx1000 committed:
Revise the readme per the requested changes.
add the screenshot.
address the comments for.

asuessenbach commented on shaders/mesh_shader_culling/mesh_shader_shared.h:
```diff
@@ -16,13 +16,12 @@
-	vec4  positionTransformation;
-	int   N;
-	int   meshletsNumber;
+	vec2  positionTransformation;
```
When you change from vec4 to vec2, you should adjust the usage in the task and mesh shader!

asuessenbach reviewed (commented):
After adjusting the task and mesh shader to the positionTransformation being a vec2 now, I get a different image at startup: the blocks are much larger and shifted to the lower right, color varies per column only, and I have to reduce the cull radius to see any culling at all.
Seems, you have to fine-adjust the settings here.

gpx1000 committed:
missing changes from previous commit.

gpx1000 committed:
change initial settings so the sample shows something correct on first run.

asuessenbach commented on shaders/mesh_shader_culling/mesh_shader_culling.mesh:
```diff
@@ -51,7 +51,8 @@ void main()
-    vec4 position = displacement + sharedData.positionTransformation;
+    vec4 posTrans = vec4(sharedData.positionTransformation.x, sharedData.positionTransformation.y, 0.0, 0.0);
+    vec4 position = displacement + posTrans;
     float squareRadius = position.x * position.x + position.y * position.y;
```
It's a bit surprising, that you extend `deltaPosition_xy` and `sharedData.positionTransform` from `vec2` to `vec4`, and then just use `x` and `y`.
Why not just
`vec2 position = deltaPosition_xy + sharedData.positionTransformation;`

gpx1000 replied:
:) Did it without thinking.  Good catch

asuessenbach commented on shaders/mesh_shader_culling/mesh_shader_culling.mesh:
```diff
@@ -68,10 +69,10 @@ void main()
+        gl_MeshVerticesEXT[0].gl_Position = vec4(2.0 * sharedData.subDimension * unitVertex_0, 0.0f, 1.0f) + posTrans + displacement;
+        gl_MeshVerticesEXT[1].gl_Position = vec4(2.0 * sharedData.subDimension * unitVertex_1, 0.0f, 1.0f) + posTrans + displacement;
+        gl_MeshVerticesEXT[2].gl_Position = vec4(2.0 * sharedData.subDimension * unitVertex_2, 0.0f, 1.0f) + posTrans + displacement;
+        gl_MeshVerticesEXT[3].gl_Position = vec4(2.0 * sharedData.subDimension * unitVertex_3, 0.0f, 1.0f) + posTrans + displacement;
```
Here, `posTrans + displacement` is the above calculated `position`.
Moreover, due to the w-component, `displacement` is a point while `posTrans` is a vector. That is, you're adding two points and a vector which leads to, well, something strange, with a w-component of 2.0f. You should not do that.
If you would just use a `vec2 position`, you could calculate
`vec4(2.0f * sharedData.subDimension * unitVertex_N + position, 0.0f, 1.0f);`

asuessenbach reviewed (commented):
The blocks are nicely centered on startup now, but with the first keystroke, it jumps to a completely different position. Only from there on, movement looks reasonable.

asuessenbach commented on samples/extensions/mesh_shader_culling/CMakeLists.txt:
```diff
@@ -0,0 +1,31 @@
+        SHADER_FILES_GLSL
+        "mesh_shader_culling/mesh_shader_culling.task"
+        "mesh_shader_culling/mesh_shader_culling.mesh"
+        "mesh_shader_culling/mesh_shader_culling.frag")
```
Add `mesh_shader_shared.h` here?

gpx1000 replied:
I went ahead and did this; however, we should have a talk in the group about best practices for header files used only in shaders being included in CMake.
CMake's normal rules are that you don't include header files in your targets, they are discovered by the build tools and CMake does the right thing from there.  However, CMake can't do the right thing in shader includes because no tooling exists that will discover and inform CMake about that header file.  We might even take it so far as to have a talk about if we want to introduce a best practice of providing the shader includes with a different extension so tooling can potentially do syntax highlighting and it will be easier to write other tools like static analysis.

gpx1000 committed:
fix issues and fix the CMake warning.
put the header file in the CMakeLists.txt.

gpx1000 committed:
fix the clang format

gpx1000 committed:
Merge branch 'main' into vk_Mesh_Shader_Culling

asuessenbach commented:
@gpx1000 I've sent you another version of a task/mesh shader that might show the "power of task shaders" a bit more...
After dispatching N*N task shader workgroups with vkCmdDrawMeshTasksEXT, there are 2*2 task shader invocations per workgroup. Some culling is performed in that shader, and a variable number of mesh shader workgroups (0....4) is emitted from one invocation per task shader workgroup. There are 2*2 mesh shader invocations per workgroup, each emitting 2 triangular primitives with 4 vertices.
That is, the final drawing looks a bit different than before.
Again, feel free to use that approach to your liking.

gpx1000 committed:
accept the contribution of changes from @asuessenbach

gpx1000 commented:
> asuessenbach

Thanks for the contribution; it is now part of the sample.

gpx1000 committed:
Merge branch 'main' into vk_Mesh_Shader_Culling
